src/foundry/common/abstract/_types.d.mts

```diff
@@ -0,0 +1,246 @@
+   */
+  broadcast: boolean;
+
+  /**
+   * An array of data objects used to update existing Documents.
+   * Each update object must contain the _id of the target Document
+   */
+  updates: foundry.data.fields.SchemaField.InnerAssignmentType<T["schema"]["fields"]>[];
+
+  /**
+   * Difference each update object against current Document data and only use
+   * differential data for the update operation
+   */
+  diff?: boolean;
+
+  /**
+   * The timestamp when the operation was performed
+   */
+  modifiedTime?: number;
+
+  /**
+   * Merge objects recursively. If false, inner objects will be replaced
+   * explicitly. Use with caution!
+   */
+  recursive?: boolean;
+
+  /**
+   * Re-render Applications whose display depends on the created Documents
+   */
+  render?: boolean;
+
+  /**
+   * Block the dispatch of hooks related to this operation
+   */
+  noHook?: boolean;
+
+  /**
+   * A parent Document within which Documents are embedded
+   */
+  parent?: foundry.abstract.Document.Any;
+
+  /**
+   * A compendium collection ID which contains the Documents
+   */
+  pack: string | null;
+
+  /**
+   * A parent Document UUID provided when the parent instance is unavailable
+   */
+  parentUuid?: string | null;
+
+  /**
+   * An alias for 'updates' used internally by the server-side backend
+   */
+  _result?: (string | object)[];
+}
+
+export interface DatabaseDeleteOperation {
+  /**
+   * Whether the database operation is broadcast to other connected clients
+   */
+  broadcast: boolean;
+
+  /**
+   * An array of Document ids which should be deleted
+   */
+  ids: string[];
+
+  /**
+   * Delete all documents in the Collection, regardless of _id
+   */
+  deleteAll?: boolean;
+
+  /**
+   * The timestamp when the operation was performed
+   */
+  modifiedTime?: number;
+
+  /**
+   * Block the dispatch of hooks related to this operation
+   */
+  noHook?: boolean;
+
+  /**
+   * Re-render Applications whose display depends on the deleted Documents
+   */
+  render?: boolean;
+
+  /**
+   * A parent Document within which Documents are embedded
+   */
+  parent?: foundry.abstract.Document.Any;
+
+  /**
+   * A compendium collection ID which contains the Documents
+   */
+  pack: string | null;
+
+  /**
+   * A parent Document UUID provided when the parent instance is unavailable
+   */
+  parentUuid?: string | null;
+
+  /**
+   * An alias for 'ids' used internally by the server-side backend
+   */
+  _result?: (string | object)[];
+}
+
+export type DatabaseAction = "get" | "create" | "update" | "delete";
+
+export type DatabaseOperation =
+  | DatabaseGetOperation
+  | DatabaseCreateOperation
+  | DatabaseUpdateOperation
+  | DatabaseDeleteOperation;
+
+export interface DocumentSocketRequest {
+  /**
+   * The type of Document being transacted
+   */
+  type: string;
+
+  /**
+   * The action of the request
+   */
+  action: DatabaseAction;
+
+  /**
+   * Operation parameters for the request
+   */
+  operation: DatabaseOperation;
```

Comment:
I assume the `action` and the `operation` needs to correspond? If so it should probably be organized like so:
```ts
export type DatabaseOperationMap = {
    get: DatabaseGetOperation;
    create: DatabaseCreateOperation;
    update: DatabaseUpdateOperation;
    delete: DatabaseDeleteOperation;
}

export type DatabaseAction = keyof DatabaseOperationMap;
export type DatabaseOperation = DatabaseOperationMap[keyof DatabaseOperationMap];

export interface DocumentSocketRequest<Action extends DatabaseAction> {
    ...
    action: Action;
    operation: DatabaseOperationMap[Action]
    ...
}
```

Reply:
Seems like it. The interface is only used in private methods so it's not really possible to test.